src/bp/core/services/converse.ts

```diff
@@ -0,0 +1,137 @@
+import { UserRepository } from 'core/repositories'
+import { TYPES } from 'core/types'
+import { InvalidParameterError } from 'errors'
+import { EventEmitter2 } from 'eventemitter2'
+import { inject, injectable, postConstruct } from 'inversify'
+import _ from 'lodash'
+
+import { Event } from '../sdk/impl'
+
+import { EventEngine } from './middleware/event-engine'
+
+export const converseApiEvents = new EventEmitter2()
+
+@injectable()
+export class ConverseService {
+  jsonMap: Map<string, object> = new Map()
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.UserRepository) private userRepository: UserRepository
+  ) {}
+
+  @postConstruct()
+  init() {
+    this.eventEngine.register({
+      name: 'converse.capture.payload',
+      description: 'Captures the response payload for the Converse API',
+      order: 10000,
+      direction: 'outgoing',
+      handler: (event, next) => {
+        this._handleCapturePayload(event)
+        next()
+      }
+    })
+    this.eventEngine.register({
+      name: 'converse.capture.nlu',
+      description: 'Captures the nlu output for the Converse API',
+      order: 10000,
+      direction: 'incoming',
+      handler: (event, next) => {
+        this._handleCaptureNlu(event)
+        next()
+      }
+    })
+  }
+
+  public async sendMessage(botId: string, userId: string, payload): Promise<any> {
+    if (!payload.text || !_.isString(payload.text) || payload.text.length > 360) {
+      throw new InvalidParameterError('Text must be a valid string of less than 360 chars')
+    }
+
+    const sanitizedPayload = _.pick(payload, ['text', 'type', 'data', 'raw'])
+
+    // We remove the password from the persisted messages for security reasons
+    if (payload.type === 'login_prompt') {
+      sanitizedPayload.data = _.omit(sanitizedPayload.data, ['password'])
+    }
+
+    if (payload.type === 'form') {
+      sanitizedPayload.data.formId = payload.formId
+    }
+
+    const { result: user } = await this.userRepository.getOrCreate('api', userId)
+
+    const incomingEvent = Event({
+      type: 'text',
+      channel: 'api',
+      direction: 'incoming',
+      payload,
+      target: userId,
+      botId
+    })
+
+    await this.eventEngine.sendEvent(incomingEvent)
+
+    const timeoutPromise = this._createTimeoutPromise(userId)
```

Comment:
potential race condition here, i.e. your event might get processed before it was listened to. better listen before awaiting for event to be sent